File: go-version/leetcode/LC3355/is_zero_array.go
package LC3355

/**
 * @Author: cutejiuge cutejiuge@163.com
 * @Date: 2025/5/20 下午11:30
 * @FilePath: is_zero_array
 * @Description:
 */

func isZeroArray(nums []int, queries [][]int) bool {
	n := len(nums)
	diffArray := make([]int, n)
	for _, query := range queries {
		diffArray[query[0]]++
		if query[1] < n-1 {
			diffArray[query[1]+1]--
		}
	}

	sumDiff := 0
	for i := 0; i < n; i++ {
		sumDiff += diffArray[i]
		if nums[i] > sumDiff {
			return false
		}
	}
	return true
}

File: python-version/LC3355/is_zero_array.py
class Solution:
    @staticmethod
    def isZeroArray(nums: list[int], queries: list[list[int]]) -> bool:
        n = len(nums)
        diff_arr = [0] * n
        for query in queries:
            diff_arr[query[0]] += 1
            if query[1] < n - 1:
                diff_arr[query[1] + 1] -= 1

        sum_diff = 0
        for i, _ in enumerate(diff_arr):
            sum_diff += diff_arr[i]
            if nums[i] > sum_diff:
                return False
        return True
